### PR TITLE
Modernize private copy ctor and assignment by using =delete

### LINKED ISSUE #1218: Use the delete keyword to forbid copy constructor and assignment
The old way to forbid copy constructors and assignment operators was to declare them private, which is used at several places in casacore. C++-11  added the delete keyword to achieve this goal. 
Change all occasions where the old way is used.

### PR DESCRIPTION
**Pull request by gervandiepen**
Several classes used the old style (making them private) to forbid use of copy constructor and assignment. They now use the modern =delete clause. They have been found using clang-tidy.
Furthermore, several clang-tidy warnings have been fixed. Most of them were about using virtual functions in constructor or destructor.

Close #1218

### CODE REVIEW

File: tables/TaQL/test/tTableGram.out [hunk 1]
```diff
@@ -64,7 +64,7 @@ select ab from tTableGram_tmp.tab where ab==2**1**2 || ab==-2**-1*8/-2*3
 1 selected columns:  ab
  2
  6
-select ab,ac,af from tTableGram_tmp.tab where lower(af) == regex("v[01279]")
+select ab,ac,af from ../../TaQL/./test/tTableGram_tmp.tab where lower(af) == regex("v[01279]")
```
Comment (tammojan) [inline tables/TaQL/test/tTableGram.out:67]: This seems to be unrelated

**Review by mordante — COMMENTED** [review]
Ger for the next review it would be great to try to keep one topic per PR.
This one has several small unrelated changes, which means I really need to look closely at the entire PR, which costs a lot of extra time. For example the change to the locking code should really be a separate review. When during a bisect an error is found in this patch people need to look in this large commit to find the possible cause between the big bulk of NFC changes.

File: tables/TaQL/test/tTableGram.out (hunk 1 above)
Comment (gervandiepen) [reply tables/TaQL/test/tTableGram.out:67]: Indeed it is testing those, but I'll undo these changes in this PR.

File: casa/HDF5/HDF5Object.h [hunk 2]
```diff
@@ -76,6 +76,12 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
     // The destructor in a derived class should close the hid appropriately.
     virtual ~HDF5Object();
 
+    // Copy constructor cannot be used.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply casa/HDF5/HDF5Object.h:79]: Added why.

File: tables/DataMan/BaseMappedArrayEngine.h [hunk 3]
```diff
@@ -295,6 +295,11 @@ template<class VirtualType, class StoredType> class BaseMappedArrayEngine : publ
     BaseMappedArrayEngine
 	              (const BaseMappedArrayEngine<VirtualType, StoredType>&);
 
+    // Assignment is not needed and therefore forbidden
+    // (so it is made private and not implemented).
+    BaseMappedArrayEngine<VirtualType, StoredType>& operator=
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/BaseMappedArrayEngine.h:300]: They used to be, but apparently not anymore. I've removed them.

File: tables/DataMan/ISMBase.cc [hunk 4]
```diff
@@ -303,7 +303,7 @@ void ISMBase::makeIndex()
     if (index_p != 0) {
 	return;
     }
-    index_p = new ISMIndex (this);
+    index_p = new ISMIndex();
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/ISMBase.cc:306]: Yes, it is. That argument is removed from ISMIndex because it was not used (as clang-tidy said).

File: tables/DataMan/ISMBase.cc [hunk 5]
```diff
@@ -303,7 +303,7 @@ void ISMBase::makeIndex()
...
     AlwaysAssert (index_p != 0, AipsError);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/ISMBase.cc:307]: You're right. I've removed it as well as a few others in this file. It stems from the old days.

File: tables/TaQL/ExprGroup.h [hunk 6]
```diff
@@ -240,9 +243,6 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
     virtual MArray<String> getArrayString (const vector<TableExprId>& = vector<TableExprId>());
     // <group>
   private:
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/ExprGroup.h:245]: Thanks.

File: tables/TaQL/RecordGram.cc [hunk 7]
```diff
@@ -94,13 +94,13 @@ void RecordGram::addToken (TableExprNodeSet* ptr)
 void RecordGram::addToken (TableExprNodeSetElem* ptr)
   { addToken (ptr, RecordGram::Elem); }
 void RecordGram::deleteToken (TableExprNode* ptr)
-  { delete ptr; removeToken (ptr); }
+  { removeToken (ptr); delete ptr; }
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/RecordGram.cc:97]: Thanks to clang-tidy.

File: tables/Tables/ColumnsIndexArray.cc [hunk 8]
```diff
@@ -521,10 +521,7 @@ void ColumnsIndexArray::getArray (Vector<uChar>& result, const String& name)
     rownr_t npts = arr.nelements();
     nrel[0] = npts;
     result.resize (nrrow*npts);
-    Bool deleteIt;
-    uChar* data = result.getStorage(deleteIt);
-    objmove (data, arr.getStorage(deleteIt), npts);
-    data += npts;
+    objmove (result.data(), arr.data(), npts);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/ColumnsIndexArray.cc:524]: It should be tested in tColumnsIndexArray.cc. Maybe I'll change this one day to using std::vector.
We should do a coverage test one day.

File: tables/DataMan/MSMDirColumn.h [hunk 9]
```diff
@@ -128,15 +134,7 @@ class MSMDirColumn: public MSMColumn
...
 
-  // Forbid copy constructor.
-  MSMDirColumn (const MSMDirColumn&);
-
-  // Forbid assignment.
-  MSMDirColumn& operator= (const MSMDirColumn&);
-
-
-  // The (unique) sequence number of the column.
-  uInt seqnr_p;
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/MSMDirColumn.h:139]: Change has been undone.

File: tables/DataMan/MSMIndColumn.h [hunk 10]
```diff
@@ -137,19 +143,17 @@ class MSMIndColumn : public MSMColumn
...
   private:
-    Data (const Data&);
-    Data& operator= (const Data&);
     IPosition shape_p;
     void* data_p;
   };
   // The shape of all arrays in case it is fixed.
   IPosition fixedShape_p;
-  // The size of an array element.
-  uInt elemSize_p;
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/MSMIndColumn.h:152]: Undone

File: tables/DataMan/MSMIndColumn.h [hunk 11]
```diff
@@ -137,19 +143,17 @@ class MSMIndColumn : public MSMColumn
     Data (const IPosition& shape, int dtype, int elemSize);
     //# explicitly specify noexcept to squash compiler warning
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/MSMIndColumn.h:144]: warning: '~Data' has a non-throwing exception specification but can still throw [-Wexceptions]
I've explained it somewhat better.

File: tables/DataMan/VirtArrCol.h [hunk 12]
```diff
@@ -183,7 +183,13 @@ class VirtualArrayColumn : public VirtualArrayColumnBase
...
     virtual ~VirtualArrayColumn();
 
-    // Return the data type of the column.
+    // The object cannot be copied.
+    VirtualArrayColumn (const VirtualArrayColumn<T>&) = delete;
+
+    // The object cannot be assigned to.
+    VirtualArrayColumn<T>& operator= (const VirtualArrayColumn<T>&) = delete;
+
+  // Return the data type of the column.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/VirtArrCol.h:192]: Fixed

File: tables/DataMan/VACEngine.h [hunk 13]
```diff
@@ -151,6 +151,9 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
 //      virtual ~AVACEngine()
 //      {}
 //
+//      // Assignment is not needed and therefore forbidden.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/VACEngine.h:154]: Changed as suggested.

File: tables/DataMan/MSMIndColumn.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (mordante) [reply tables/DataMan/MSMIndColumn.h]: I see, I really dislike throwing destructors. They may cause an application to using the library to terminate. I think this code can easily be improved. I will create a patch for it.

File: tables/DataMan/BaseMappedArrayEngine.h [hunk 14]
```diff
@@ -290,10 +290,13 @@ template<class VirtualType, class StoredType> class BaseMappedArrayEngine : publ
...
-    // (so it is made protected).
-    BaseMappedArrayEngine
-	              (const BaseMappedArrayEngine<VirtualType, StoredType>&);
+    // Copy constructor is used by copy constructor of derived classes.
+    BaseMappedArrayEngine (const BaseMappedArrayEngine&);
+
+    // Assignment is not needed and therefore forbidden
+  ///    BaseMappedArrayEngine<VirtualType, StoredType>& operator=
+  //	             (const BaseMappedArrayEngine<VirtualType, StoredType>&) = delete;
+    BaseMappedArrayEngine& operator= (const BaseMappedArrayEngine&) = delete;
```
Comment (mordante) [inline tables/DataMan/BaseMappedArrayEngine.h:299]: Some leftover comment.

**Review by mordante — APPROVED** [review]
LGTM modulo one comment.
